File: LinkedList/java/leetcode148_Sort_List.java
//Sort a linked list in O(n log n) time using constant space complexity.
//
//Example 1:
//
//Input: 4->2->1->3
//Output: 1->2->3->4
//Example 2:
//
//Input: -1->5->3->4->0
//Output: -1->0->3->4->5

package wenxinjie2;

public class leetcode148_Sort_List {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode mid = getMid(head);
        ListNode head2 = mid.next;
        mid.next = null; 
        return merge(sortList(head), sortList(head2));       
    }
    
    public ListNode getMid(ListNode head){
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null && fast.next.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
    
    public ListNode merge(ListNode l, ListNode r){
        if (l.val >= r.val){
            ListNode tem = l; l = r; r = tem;
        }
        ListNode p = l;
        while (l != null && l.next != null && r != null){
            if (l.val <= r.val && r.val <= l.next.val){
                ListNode tem1 = r.next;
                r.next = l.next;
                l.next = r;
                r = tem1;
                l = l.next;                   
            }
            else if (r.val > l.next.val) l = l.next;
        }
        if (l.next == null) l.next = r;
        return p;
    }
}

//Time: O(nlog(n))
//Space: O(1) 
//Difficulty: medium

File: LinkedList/python/leetcode148_Sort_List.py
# Sort a linked list in O(n log n) time using constant space complexity.

# Example 1:

# Input: 4->2->1->3
# Output: 1->2->3->4
# Example 2:

# Input: -1->5->3->4->0
# Output: -1->0->3->4->5

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head
        mid = self.getmid(head)
        head2 = mid.next
        mid.next = None
        
        return self.merge(self.sortList(head), self.sortList(head2))
    
    def getmid(self, head):
        fast = head
        slow = head
        while fast and fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        return slow
    
    def merge(self, l, r):
        if l.val >= r.val: 
            l, r = r, l
        p = l
        while l and r and l.next:
            if  l.val <= r.val <= l.next.val:
                tem = r.next
                r.next = l.next
                l.next = r
                r = tem
                l = l.next
            elif r.val > l.next.val:
                l = l.next
        if l.next == None:
            l.next = r        
        return p

# Time: O(nlog(n))
# Space: O(1) 
# Difficulty: medium